### PR TITLE
Combine error types

### LINKED ISSUE #100: Combine SerializeError and DeserializeError, SerializeResult and DeserializeResult
See https://github.com/TyOverby/bincode/pull/95#issuecomment-276457709.

### PR DESCRIPTION
**Pull request by dgriffen**
Closes #100

### CODE REVIEW

**Comment by dtolnay** [comment]
I can recommend what I have done for `serde::de::value::Error`, `serde_json::Error` and `serde_yaml::Error` which is hide them behind an opaque newtype wrapper.

```rust
pub struct Error(Box<ErrorImpl>);

enum ErrorImpl {
    /* variants */
}
```

The Box is to drop the size of Error from 48 bytes down to 8 bytes which sped up performance of all of serde_json by 5%, and the newtype wrapper is to let us add or change error variants without a breaking change. Nobody actually pattern-matches on these errors - they just print the message and fail.

**Comment by dgriffen** [comment]
Sounds like a good idea. @TyOverby what do you think?

**Comment by TyOverby** [comment]
@dtolnay @dgriffen: I'm all for sticking it inside of a Box, but my use-cases for bincode require pattern matching on the error: I *really* care if the deserialization failed because of a dropped network connection (IoError) or because they sent bad bytes over the wire (programmer error).

**Comment by TyOverby** [comment]
I am surprised that allocating is cheaper than moving an extra 40 bytes though.

**Comment by TyOverby** [comment]
Oh, I guess if you rarely hit error cases, then it's a pretty good speedup in the general case.

**Comment by dgriffen** [comment]
Oops, looks like I made a mistake when fixing conflicts. That's the last time I use the Github's conflict resolver.

**Comment by dgriffen** [comment]
So how should I move forward with any changes to `Error`? It sounds like we either change the `Result` type to carry a `Box<Error>` or do what @dtolnay said, but make `ErrorImpl` public and give it a better name if we want pattern matching.

**Comment by TyOverby** [comment]
@dgriffen: yeah, that sounds like a good move forward!  Off the top of my head I can't think of a good enum name though.  `ErrorKind` maybe?

**Comment by dgriffen** [comment]
`ErrorKind` sounds good to me.

**Comment by TyOverby** [comment]
I tried to resolve the merge conflicts with githubs online tool.  Let's see if it actually worked!

**Comment by dgriffen** [comment]
Looks like you kept the old error types

**Comment by TyOverby** [comment]
Well, that's the last time I use that tool.

**Comment by dgriffen** [comment]
https://github.com/TyOverby/bincode/pull/104#issuecomment-276570084

**Comment by TyOverby** [comment]
And here I thought I was helping :|

**Comment by dgriffen** [comment]
Looks like a few more merge errors that didnt get caught, I'll clean those up

**Comment by TyOverby** [comment]
@dtolnay Want to do another pass?

File: src/serde/mod.rs [hunk 1]
```diff
@@ -25,15 +22,115 @@ use serde_crate as serde;
...
+    SequenceMustHaveLength,
+    Custom(String)
+}
+
+impl error::Error for ErrorKind {
+    fn description(&self) -> &str {
+        match *self {
+            ErrorKind::IoError(ref err) => error::Error::description(err),
+            ErrorKind::InvalidEncoding(ref ib) => ib.desc,
+            ErrorKind::SequenceMustHaveLength => "bincode can't encode infinite sequences",
```
Comment (dtolnay) [inline src/serde/mod.rs]: I don't think this error is accurate - the one you have in Display is a better characterization of SequenceMustHaveLength.

File: src/serde/mod.rs [hunk 2]
```diff
@@ -25,15 +22,115 @@ use serde_crate as serde;
...
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ErrorKind::IoError(ref ioerr) =>
+                write!(fmt, "IoError: {}", ioerr),
+            ErrorKind::InvalidEncoding(ref ib) =>
+                write!(fmt, "InvalidEncoding: {}", ib),
+            ErrorKind::SequenceMustHaveLength =>
+                write!(fmt, "Bincode can only encode sequences and maps that have a knowable size ahead of time."),
+            ErrorKind::SizeLimit =>
+                write!(fmt, "SizeLimit"),
```
Comment (dtolnay) [inline src/serde/mod.rs]: This is the Display representation not the Debug representation, so please make this message self-explanatory.

File: src/serde/mod.rs [hunk 3]
```diff
@@ -25,15 +22,115 @@ use serde_crate as serde;
...
+    fn from(err: IoError) -> Error {
+        ErrorKind::IoError(err).into()
+    }
+}
+
+impl fmt::Display for ErrorKind {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ErrorKind::IoError(ref ioerr) =>
+                write!(fmt, "IoError: {}", ioerr),
```
Comment (dtolnay) [inline src/serde/mod.rs]: Rust identifiers like `IoError` belong in the Debug representation. For the Display representation I would prefer "IO error: {}".

File: src/serde/mod.rs [hunk 4]
```diff
@@ -25,15 +22,115 @@ use serde_crate as serde;
...
+            ErrorKind::InvalidEncoding(ref ib) =>
+                write!(fmt, "InvalidEncoding: {}", ib),
```
Comment (dtolnay) [inline src/serde/mod.rs]: Ditto here - "invalid encoding: {}".

**Comment by dtolnay** [comment]
The code looks good but the error messages are not as helpful as they could be. Ok to merge and follow up under #98 in another PR.

**Comment by TyOverby** [comment]
Looks good!  We'll take care of error messages as a part of #98

**Pull request merged**